synapse/rest/media/v1/preview_url_resource.py

```diff
@@ -329,20 +331,74 @@ def _calc_og(self, tree, media_info, requester):
                 # ...or if they are within a <script/> or <style/> tag.
                 # This is a very very very coarse approximation to a plain text
                 # render of the page.
-                text_nodes = tree.xpath("//text()[not(ancestor::header | ancestor::nav | "
-                                        "ancestor::aside | ancestor::footer | "
-                                        "ancestor::script | ancestor::style)]" +
-                                        "[ancestor::body]")
-                text = ''
+
+                # We don't just use XPATH here as that is slow on some machines.
+
+                cloned_tree = deepcopy(tree.find("body"))
```

Comment:
I felt it rude to completely destroy the only copy of tree (by removing lots of bits)